File: mobile-sim/java/src/soargroup/rosie/perception/ObjectProperty.java
package soargroup.rosie.perception;

import java.util.*;
import java.util.HashMap;

import sml.*;
import edu.umich.rosie.soar.FloatWME;
import edu.umich.rosie.soar.ISoarObject;

/**
 * A category for each object, which contains several possible labels and their confidences
 * 
  * 
 */
public class ObjectProperty implements ISoarObject
{   
    protected String name;
	protected HashMap<String, Double> lastUpdate;
    protected HashMap<String, FloatWME> values;
    
    public ObjectProperty(String propName){
		name = propName;
		lastUpdate = new HashMap<String, Double>();
		values = new HashMap<String, FloatWME>();
    }
    
    public String getPropertyName(){
    	return name;
    }

	public void clearValues(){
		lastUpdate.clear();
	}

	public void addValue(String value, Double confidence){
		lastUpdate.put(value, confidence);
	}

    /**************************************************************
     * Methods for adding to working memory
     **************************************************************/

	protected Identifier propId = null;
    protected Identifier valuesId = null;
    
    public boolean isAdded(){
		return (propId != null);
    }
    
    public void addToWM(Identifier parentId){
		if (propId != null){ return; }

		propId = parentId.CreateIdWME("property");
		propId.CreateStringWME("property-handle", name);

		valuesId = propId.CreateIdWME("values");
		for(Map.Entry<String, Double> e : lastUpdate.entrySet()){
			FloatWME wme = new FloatWME(e.getKey(), e.getValue());
			wme.addToWM(valuesId);
			values.put(e.getKey(), wme);
		}
    }
	
	public void updateWM(){
		if(propId == null){ return; }

		for(Map.Entry<String, Double> e : lastUpdate.entrySet()){
			FloatWME wme = values.get(e.getKey());
			if (wme == null){
				wme = new FloatWME(e.getKey(), e.getValue());
				wme.addToWM(valuesId);
				values.put(e.getKey(), wme);
			} else if(Math.abs(wme.getValue() - e.getValue()) > 0.02){
				wme.setValue(e.getValue());
				wme.updateWM();
			}
		}

    	HashSet<String> oldVals = new HashSet<String>();
    	oldVals.addAll(values.keySet());
		oldVals.removeAll(lastUpdate.keySet());
		for(String val : oldVals){
			values.get(val).removeFromWM();
			values.remove(val);
		}
    }
    
    public void removeFromWM(){
		if(propId != null){ return; }

    	for(FloatWME wme : values.values()){
    		wme.removeFromWM();
    	}
		values.clear();
    	valuesId.DestroyWME();
    	valuesId = null;

    	propId.DestroyWME();
    	propId = null;
    }
}

File: mobile-sim/python/mobilesim/rosie/perception/ObjectProperty.py
from pysoarlib import *


class ObjectProperty(WMInterface):
    def __init__(self, name):
        super().__init__()
        self.name = name
        self.root_id = None
        self.values_id = None

        self.last_update = {}
        self.values = {}
    
    def get_name(self):
        return self.name

    def clear_values(self):
        self.last_update = {}
    
    def add_value(self, value, conf):
        self.last_update[value] = conf

    def get_best_value(self):
        best_conf = 0.0
        best_val = None
        for val, conf in self.last_update.items():
            if conf > best_conf:
                best_conf = conf
                best_val = val
        return best_val

    ### Methods for managing working memory structures ###

    def _add_to_wm_impl(self, parent_id):
        self.root_id = parent_id.CreateIdWME("property")
        self.root_id.CreateStringWME("property-handle", self.name)

        self.values_id = self.root_id.CreateIdWME("values")
        for val, conf in self.last_update.items():
            self.values[val] = SoarWME(val, conf)
            self.values[val].add_to_wm(self.values_id)
        
    def _update_wm_impl(self):
        for val, conf in self.last_update.items():
            wme = self.values.get(val, None)
            if wme is None:
                wme = SoarWME(val, conf)
                wme.add_to_wm(self.values_id)
                self.values[val] = wme
            elif abs(wme.get_value() - conf) > 0.02:
                wme.set_value(conf)
                wme.update_wm()

        old_vals = [ val for val in self.values if val not in self.last_update ]
        for val in old_vals:
            self.values[val].remove_from_wm()
            del self.values[val]

    def _remove_from_wm_impl(self):
        for val, wme in self.values.items():
            wme.remove_from_wm()
        self.root_id.DestroyWME()
        self.root_id = None
        self.values_id = None
